numpy/core/src/umath/ufunc_object.c

```diff
@@ -1731,9 +1746,145 @@ make_arr_prep_args(npy_intp nin, PyObject *args, PyObject *kwds)
+                int broadcast_ndim, int do_remap[NPY_MAXARGS],
+                int remap_axis[NPY_MAXARGS][NPY_MAXDIMS]) {
+    int nin = ufunc->nin;
+    int nout = ufunc->nout;
+    int nop = nin + nout;
+    int op_ndim, core_ndim, num_axes;
+    int i, j, iter_axis;
+    int requested_axis = 0;
+    int list_size = 0;
+    int item_is_tuple = 0;
+    PyObject *item = NULL;
+    int have_seen_axis[NPY_MAXDIMS];
+
+    if (!PyList_Check(axes)) {
+        PyErr_SetString(PyExc_ValueError, "axes should be a list");
+        return -1;
+    }
+    list_size = PyList_Size(axes);
+    if(list_size != nop) {
+        int size_ok = (list_size == nin);
+
+        if (size_ok) {
+            /* we allow not giving axes for outputs without core dimensions */
+            for (i = list_size; i < nop; ++i) {
+                do_remap[i] = 0;
+                if(ufunc->core_num_dims[i] > 0) {
+                    size_ok = 0;
+                    break;
+                }
+            }
+        }
+        if(!size_ok) {
+            PyErr_Format(PyExc_ValueError,
+                         "axes should be a list with an entry for all "
+                         "%d inputs and outputs; entries for outputs can "
+                         "only be omitted if they do not have core axes.",
+                         nop);
+            return -1;
+        }
+    }
+    for (i = 0; i < list_size; ++i) {
+        core_ndim = ufunc->core_num_dims[i];
+        if (op[i] != NULL) {
+            op_ndim = PyArray_NDIM(op[i]);
+        }
+        else {
+            op_ndim = broadcast_ndim + core_ndim;
+        }
+        item = PyList_GET_ITEM(axes, i);
+        item_is_tuple = PyTuple_Check(item);
+        if (item_is_tuple) {
+            num_axes = PyTuple_Size(item);
+        }
+        else {
+            /* try as integer */
+            requested_axis = (int)PyInt_AsSsize_t(item);
+            if (error_converting(requested_axis)) {
+                PyErr_Clear();
+                /* to give descriptive error message below */
+                num_axes = -1;
+            }
+            else {
+                num_axes = 1;
+            }
+        }
+        if (num_axes != core_ndim) {
+            if (core_ndim == 1) {
+                PyErr_Format(PyExc_ValueError,
+                             "axes item %d should be a tuple with a "
+                             "single element, an integer, or None", i);
+            }
+            else {
+                PyErr_Format(PyExc_ValueError,
+                             "axes item %d should be a tuple with %d "
+                             "elements, or None", i, core_ndim);
+            }
+            return -1;
+        }
+        /*
+         * Now create the remap, starting with the core dimensions, and
+         * then adding the remaining axes that are to be iterated over.
+         */
+        for (j = 0; j < op_ndim; j++) {
+            have_seen_axis[j] = 0;
```

Comment:
Declaring `have_seen_axis` inside this loop would make it even clearer that it shares no state between iterations.